File: tools/serve.py
#!/usr/bin/env python3
"""Static file server for development, with caching turned off.

`python -m http.server` sends no cache headers beyond Last-Modified, which leaves
the browser applying heuristic caching. For ES modules that is worse than it
sounds: each file caches independently, so a reload can end up running a new
index.html against an old module, and the resulting behaviour matches neither
version of the code. Debugging that costs more than the bandwidth ever saved.

The published site is served by GitHub Pages with proper ETags, so this only
affects development.

    python tools/serve.py [port]
"""

from __future__ import annotations

import functools
import http.server
import socketserver
import sys

DEFAULT_PORT = 8643

# Types the shipped pages fetch that Python's own table gets wrong or misses.
# .glb.gz in particular: the site fetches it and inflates with DecompressionStream,
# so it must arrive as an opaque body, not something the browser tries to decode.
EXTRA_TYPES = {
    ".glb": "model/gltf-binary",
    ".gz": "application/octet-stream",
    ".webp": "image/webp",
    ".wasm": "application/wasm",
    ".mjs": "text/javascript",
    ".js": "text/javascript",
    ".json": "application/json",
    ".hca": "application/octet-stream",
    ".awb": "application/octet-stream",
    ".acb": "application/octet-stream",
}


class NoCacheHandler(http.server.SimpleHTTPRequestHandler):
    extensions_map = {
        **http.server.SimpleHTTPRequestHandler.extensions_map,
        **EXTRA_TYPES,
    }

    def end_headers(self):
        self.send_header("Cache-Control", "no-store, must-revalidate")
        self.send_header("Pragma", "no-cache")
        self.send_header("Expires", "0")
        super().end_headers()

    def send_header(self, keyword, value):
        # Last-Modified is what the browser bases heuristic caching on, and
        # no-store alone does not always stop a module from being reused.
        if keyword == "Last-Modified":
            return
        super().send_header(keyword, value)

    def log_message(self, fmt, *args):
        # One line per asset is thousands of lines for a single page load here.
        # Errors still print, which is the part worth seeing.
        status = args[1] if len(args) > 1 else ""
        if str(status).startswith(("4", "5")):
            super().log_message(fmt, *args)


class Server(socketserver.ThreadingTCPServer):
    # Without this a restart on the same port fails with "address already in use"
    # for as long as the old socket sits in TIME_WAIT.
    allow_reuse_address = True
    daemon_threads = True


def main() -> int:
    port = int(sys.argv[1]) if len(sys.argv) > 1 else DEFAULT_PORT
    handler = functools.partial(NoCacheHandler, directory=".")
    with Server(("", port), handler) as httpd:
        print(f"serving . on http://localhost:{port}  (no-store)")
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            print("\nstopped")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tools/serve.mjs
#!/usr/bin/env node
/**
 * Node equivalent of tools/serve.py — a static file server with caching
 * turned off, so a reload never runs a new index.html against an old ES
 * module. The published site is served by GitHub Pages with proper ETags,
 * so this only affects development.
 *
 *     node tools/serve.mjs [port]
 */

import http from "node:http";
import { promises as fs } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = path.dirname(path.dirname(fileURLToPath(import.meta.url)));
const PORT = Number(process.argv[2]) || 8643;

// Types the shipped pages fetch that Node has no table for at all.
// .glb.gz in particular: the site fetches it and inflates with
// DecompressionStream, so it must arrive as an opaque body, not something
// the browser tries to decode.
const TYPES = {
  ".html": "text/html; charset=utf-8",
  ".css": "text/css; charset=utf-8",
  ".js": "text/javascript",
  ".mjs": "text/javascript",
  ".json": "application/json",
  ".glb": "model/gltf-binary",
  ".gz": "application/octet-stream",
  ".webp": "image/webp",
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".svg": "image/svg+xml",
  ".ico": "image/x-icon",
  ".wasm": "application/wasm",
  ".mp3": "audio/mpeg",
  ".hca": "application/octet-stream",
  ".awb": "application/octet-stream",
  ".acb": "application/octet-stream",
  ".txt": "text/plain; charset=utf-8",
  ".md": "text/markdown; charset=utf-8",
};

const server = http.createServer(async (req, res) => {
  const fail = (code, text) => {
    res.writeHead(code, { "Content-Type": "text/plain; charset=utf-8" });
    res.end(text);
  };

  try {
    let urlPath;
    try {
      urlPath = decodeURIComponent(new URL(req.url, "http://localhost").pathname);
    } catch {
      return fail(400, "bad request");
    }
    if (urlPath.endsWith("/")) urlPath += "index.html";

    // Resolve inside ROOT only; encode each segment back so %-tricks or ".."
    // cannot escape the site directory.
    const rel = path.normalize(urlPath).replace(/^([/\\])+/, "");
    const file = path.join(ROOT, rel);
    if (!file.startsWith(ROOT + path.sep) && file !== ROOT) return fail(403, "forbidden");

    const stat = await fs.stat(file).catch(() => null);
    if (!stat || !stat.isFile()) {
      console.error(`404 ${urlPath}`);
      return fail(404, "not found");
    }

    res.writeHead(200, {
      "Content-Type": TYPES[path.extname(file).toLowerCase()] || "application/octet-stream",
      "Content-Length": stat.size,
      // No-store + no Last-Modified: the browser must not reuse modules.
      "Cache-Control": "no-store, must-revalidate",
      Pragma: "no-cache",
      Expires: "0",
    });
    (await fs.open(file, "r")).createReadStream().pipe(res);
  } catch (err) {
    console.error(err);
    fail(500, "internal error");
  }
});

server.listen(PORT, () => {
  console.log(`serving ${ROOT} on http://localhost:${PORT}  (no-store)`);
});
